### ISSUE TITLE
[SDK 38] getExpoPushTokenAsync() throws error in android production

### ISSUE THREAD

**Issue by JamieS1211** [expo/expo#9061]
### Summary of Issue (just a few sentences)
When building to standalone android APK, the "getExpoPushTokenAsync()" function from "expo-notifications" throws an error relating to a Firebase token fetch issue despite google-services.json file being present and referenced in app.json setup.

### Environment - output of `expo diagnostics` & the platform(s) you're targeting
Expo CLI 3.21.13 environment info:
    System:
      OS: Windows 10 10.0.18362
    Binaries:
      Node: 12.16.0 - C:\Program Files\nodejs\node.EXE
      npm: 6.13.7 - C:\Program Files\nodejs\npm.CMD   

Android standalone app running on android 9 security patch 1 Oct 2019 (OxygenOS 9.0.6)

### Reproducible Demo

https://snack.expo.io/@jamies1211/expopushtoken-collection

### Steps to Reproduce

1.  Using code as shown in snack
2. Configure FCM as outlined in this guide - https://docs.expo.io/push-notifications/using-fcm/
3. Build standalone android app

### Expected Behavior vs Actual Behavior
"getExpoPushTokenAsync()" throws error rather than resolving ExpoPushToken object required to send android devices push notifications.
![Capture](https://user-images.githubusercontent.com/8209033/86327172-bcce8e80-bc3a-11ea-9b83-5dc6300488f2.PNG)

**Comment by jamesisaac** [comment]
Yup, I'm getting the same error in a standalone app (except with `getDevicePushTokenAsync` rather than `getExpoPushTokenAsync`, because I want to interact with Firebase directly).  Followed all the steps [here](https://docs.expo.io/push-notifications/using-fcm/) for enabling FCM.

To help those searching for this issue, the error message is:

> Fetching the token failed: FIS_AUTH_ERROR

**Comment by JamieS1211** [comment]
> Yup, I'm getting the same error in a standalone app (except with `getDevicePushTokenAsync` rather than `getExpoPushTokenAsync`, because I want to interact with Firebase directly). Followed all the steps [here](https://docs.expo.io/push-notifications/using-fcm/) for enabling FCM.
> 
> To help those searching for this issue, the error message is:
> 
> > Fetching the token failed: FIS_AUTH_ERROR

I am still struggling with this issue, would you mind reaching out if you make any headway and I will do likewise :)

**Comment by cruzach** [comment]
@JamieS1211 the snack you shared doesn't run 🙁 

I tested this myself on SDK 38 with a standalone app on Android 9, and getting the push token didn't cause any problems, and I was able to successfully send/receive push notifications. I used the code from the [docs demo](https://docs.expo.io/versions/latest/sdk/notifications/) and here is the relevant portion of app.json:
```
"expo": {
    "android": {
      "package": "com.example.mobile",
      "useNextNotificationsApi": true,
      "googleServicesFile": "./google-services.json"
    },
}
```

Can either of you try building a totally brand new `expo init` app (make sure `expo-cli` is up to date), using that snack's code, and configuring FCM?

**Comment by skhanker** [comment]
Was following this as I had a smattering of Push Notifications issues after upgrading a project from sdk37 to sdk38 (testing standalone apk builds).

After upgrading from expo-cli v3.21.13 to v3.22.0, creating a new sdk38 project and testing this out, it worked as expected. Slowly porting over my old project code into this new one and testing builds, it hasn't broken again.

**Comment by blackjacksme** [comment]
i've tried building new `expo init` app with updated sdk and all, then imported my project.
There is no problem getting the token when i used the expo app, but i still keep getting this error in android standalone builds

`Fetching the token failed: FIS_AUTH_ERROR`

**Comment by mczernek** [comment]
Do those experiencing issues have this line added in their `app.json`, as @cruzach mentioned?
```
"useNextNotificationsApi": true,
```
Next notifications API is the one we are willing to be the only one in the near future, so out focus is on fixing issues in it.

**Comment by blackjacksme** [comment]
i did have that line in my app.json

**Comment by jamesisaac** [comment]
@mczernek Yup, this is with `"useNextNotificationsApi": true,`, and following all the instructions regarding the latest API.

@cruzach Just created a new project with `expo init` on latest CLI, made sure I had the 3 app.json options you highlighted configured, and tested the code from the Snack.  It works when running within Expo, but once published as a standalone app it no longer works, alerting about the `FIS_AUTH_ERROR` / `E_REGISTRATION_FAILED`.  This is on Android 10.

When you tested it, was that as a standalone app?  I've not had any issues getting the token from within the Expo host... it's only in standalone where things stop working (like @blackjacksme says too).

**Comment by AybimYazilim** [comment]
I faced same problem after upgrade to sdk38. But alerting about 

> "Fetching the token failed.Given string is empty or null", code: "E_REGISTRATION_FAILED".


 I added `"useNextNotificationsApi": true`  line to app.json file. Problem still exist.

**Comment by roberts98** [comment]
@AybimYazilim I had the same error, but I didn't have googleServicesFile specified in app.json, do you have it?
Btw I didn't need the `"useNextNotificationsApi": true`

**Comment by JoaRiski** [comment]
Opened a somewhat similar issue #9061, which might not be directly related, but dropping it here in case someone ends up to this thread from Google for example

**Comment by mczernek** [comment]
As @JoaRiski notices in [#9347](https://github.com/expo/expo/issues/9347), it is a configuration issue.

We'll rethink out documentation on that, but as of now, we close this issue.

Please refer to following documentation sites:
* [Push Notifications Setup](https://docs.expo.io/push-notifications/push-notifications-setup/)
* [Using FCM](https://docs.expo.io/push-notifications/using-fcm/)

**Comment by jamesisaac** [comment]
@mczernek That's a different error though.  I was also able to recreate the `Fetching the token failed: Given String is empty or null` error in a situation where the JS was using notifications but the `app.json` wasn't fully configured.

But the error reported in the first post, `Fetching the token failed: FIS_AUTH_ERROR`, is what I see after downloading and linking a `google-services.json` file in `app.json`, and following every other step in the FCM guide you linked, including making sure the package names match.

Here's the full contents of `app.json` and `google-services.json` (with exact IDs replaced with xxxxxx)  from the root of an `expo init` project.  JS code is as in the Snack from the first post.  Would you be able to point out what the configuration issue is in this instance:

<details>
  <summary>app.json - Click to expand</summary>

```json
{
  "expo": {
    "name": "PushNotificationsTest",
    "slug": "PushNotificationsTest",
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "android": {
      "package": "com.pushnotificationtest",
      "googleServicesFile": "./google-services.json",
      "useNextNotificationsApi": true
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true
    },
    "web": {
      "favicon": "./assets/favicon.png"
    }
  }
}
```
</details>


<details>
  <summary>google-services.json - Click to expand</summary>

```json
{
  "project_info": {
    "project_number": "xxxxxx",
    "firebase_url": "https://xxxxxx.firebaseio.com",
    "project_id": "xxxxxx",
    "storage_bucket": "xxxxxx.appspot.com"
  },
  "client": [
    {
      "client_info": {
        "mobilesdk_app_id": "1:xxxxxx:android:xxxxxx",
        "android_client_info": {
          "package_name": "com.pushnotificationtest"
        }
      },
      "oauth_client": [
        {
          "client_id": "xxxxxx.apps.googleusercontent.com",
          "client_type": 3
        }
      ],
      "api_key": [
        {
          "current_key": "xxxxxx"
        }
      ],
      "services": {
        "appinvite_service": {
          "other_platform_oauth_client": [
            {
              "client_id": "xxxxxx.apps.googleusercontent.com",
              "client_type": 3
            }
          ]
        }
      }
    }
  ],
  "configuration_version": "1"
}

```

</details>

**Comment by mczernek** [comment]
@jamesisaac I've just tested it with proper and working google-services.json and it works fine. The error occurred when I had wrongly set package in `google-services.json`.

Could you please show how your repo is structured? Do you see that your app registered to FCM in FCM console?

**Comment by jamesisaac** [comment]
Just managed to track down the solution for my situation.

Prior to setting up push notifications, I'd followed the steps for setting up MapView with the Google Maps API: https://docs.expo.io/versions/latest/sdk/map-view/ , which involves setting up a Google API key with some restrictions.

When I later enabled Firebase SDK, even though it creates a new unrestricted API key, the key it used in the generated `google-services.json` was the old restricted one.

Removing these restrictions via **Google Cloud Platform -> APIs & Services -> Credentials** solved the problem (as I'm sure switching to the auto generated Firebase key would have).

Thanks for looking into it @mczernek and sorry for not spotting the key mismatch sooner.

Might be worth making these errors more specific to what could be wrong in an Expo project, or a troubleshooting section in the docs, as nearly all the Google results for `FIS_AUTH_ERROR` were talking about library version problems.  In Expo projects I suspect the situations would be:

- **`Fetching the token failed: Given String is empty or null`** -> `google-services.json` was not correctly linked in `app.json` last time the standalone app was built.
- **`Fetching the token failed: FIS_AUTH_ERROR`** -> the credentials in `google-services.json` weren't able to authenticate for Firebase Messaging; check the APIs are enabled and the key in `client.api_key.current_key` has permission to access them.

**Comment by JamieS1211** [comment]
@jamesisaac thank you so much worked a charm. Was starting to pull my hair out after trying everything mentioned here but we are all good now!

**Comment by mczernek** [comment]
Thanks a lot for the update @jamesisaac! That's a relief it works! Hopefully others will benefit from your findings!

**Comment by Akarshit** [comment]
@jamesisaac I am in the same situation as you. But I haven't ever created a restricted key(nor a un-restricted one)
Would you have any suggestions? Should I create a new unrestricted key and try adding it to the google-services.json?

**Comment by penguin-jedi** [comment]
I don't know why. Only **Firebase Cloud Messaging API** is not enough, I need **Firebase Installations API** too.

**Comment by DavidBrear** [comment]
@penguin-jedi that could make sense. I had this issue and couldn't stop getting `FIS_AUTH_ERROR` until I put the key as `Don't restrict key`. I am too scared to go back to restricting what I had been restricting before but for anyone else who's looking for a solution that is similar to mine:
- Use the `Application restrictions` to `Android Apps`
- Set the `Restrict usage to your android apps` to be the SHA-1 fingerprints for your app. (This will remove the warning of having an unrestricted key)
- Set the `API restrictions` to `Don't restrict key`

Also, thanks you @jamesisaac for your comment. I had the keys that were generated by firebase and kept trying to use them.

**Comment by AybimYazilim** [comment]
> @AybimYazilim I had the same error, but I didn't have googleServicesFile specified in app.json, do you have it?
> Btw I didn't need the `"useNextNotificationsApi": true`

Yes i have googleServicesFile in my app.json file. But it dont resolve the problem

**Comment by renanmav** [comment]
>  I'd followed the steps for setting up MapView with the Google Maps API

I'm facing the same problem, but I'm not using Google Maps API. I have configured `google-services.json` though

**Comment by impelsystems** [comment]
For anyone struggling with this, ensure you have the correct SHA1 fingerprint against the firebase registration. If you are letting Google Play manage the app signing, you can get this fingerprint from the Google Play Console > Your app > Setup > App signing > App signing key certificate > SHA-1 certificate fingerprint.

**Comment by mustafahafidi** [comment]
I manually had to create an API key in the google cloud console and enable it for FCM to get this to work (apparently the firebase provided API key is somewhat restricted).

**Comment by yoojeen126** [comment]
@jamesisaac Thanks. I just founded cloud messaging api not enabled in google cloud platform console.
I dont know why that disabled even add android package in firebase. 
But after enabled cloud messaging in google cloud platform its work fine now. Thank  you.

**Comment by ShaneMcNamara** [comment]
@jamesisaac Thank you! Looked over everything 100 times, but this was so obscure that I never considered the connection.

@mczernek, @brentvatne  - Please update the docs here (https://docs.expo.io/versions/latest/sdk/notifications/#common-gotchas--known-issues) so others don't dive down the rabbit hole on known issues.

**Comment by cruzach** [comment]
@ShaneMcNamara would you like to open up a docs PR adding this? Please feel free to add me as a reviewer 😄

**Comment by pranjal1127** [comment]
Hey @JamieS1211 
I am getting same error `Fetching the token failed: FIS_AUTH_ERROR`

I have enabled FCP API and set Credentials to GCP also.
Can You help me in this

**Comment by cruzach** [comment]
@pranjal1127 each time this happens, it's because [one of these steps](https://docs.expo.io/push-notifications/using-fcm/) was skipped, so I recommend going back over that guide

**Comment by pranjal1127** [comment]
@cruzach I have already followed these steps.But it is still giving same error

**Comment by Kubaguette** [comment]
Still doesnt work for me. I have unrestricted API key using and I still get the error `Fetching the token failed: FIS_AUTH_ERROR`. I have the correct API key and also I followed the guidelines extremely carefully. When in developer mode and with my phone connected to the pc, everything works fine but when I create a bundle and download it from the play store, the error shows always and I cant work around it.
